rclpy/rclpy/impl/logger_rcutils.py

```diff
@@ -0,0 +1,275 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from collections import OrderedDict
+try:
+    import cPickle as pickle
+except ImportError:
+    import pickle
+import importlib
+import inspect
+import time
+
+_rclpy_logging = importlib.import_module('._rclpy_logging', package='rclpy')
+
+_internal_callers = None  # List of known filenames from which logging methods can be called.
+
+
+def _find_caller(frame):
+    """Get the first calling frame that is outside of rclpy."""
+    global _internal_callers
+    if _internal_callers is None:
+        # Populate the list of known filenames from which logging methods can be called.
+        # This has to be done from within a function to avoid cyclic module imports.
+        import rclpy.logging
+        _internal_callers = [__file__, rclpy.logging.__file__]
+
+    frame_filename = inspect.getabsfile(frame)
+    while any(f in frame_filename for f in _internal_callers):
+        frame = frame.f_back
+        frame_filename = inspect.getabsfile(frame)
+    return frame
+
+
+class CallerId:
+
+    def __init__(self, frame=None):
+        if not frame:
+            frame = _find_caller(inspect.currentframe())
+        self.function_name = frame.f_code.co_name
+        self.file_name = inspect.getabsfile(frame)
+        self.line_number = frame.f_lineno
+        self.last_index = frame.f_lasti  # To distinguish between two callers on the same line
+
+
+class LoggingFilter:
+    """Base class for logging filters."""
+
+    """
+    Parameters of a filter and their default value, if appropriate.
+
+    A default value of None makes a parameter required.
+    """
+    params = {}
+
+    """
+    Initialize the context of a logging call, e.g. declare variables needed for
+    determining the log condition and add them to the context.
+    """
+    @classmethod
+    def initialize_context(cls, context, **kwargs):
+        # Store all parameters in the context so we can check that users never try to change them.
+        for param in cls.params:
+            context[param] = kwargs.get(param, cls.params[param])
+            if context[param] is None:
+                raise RuntimeError(
+                    'Required parameter "{0}" was not specified for logging filter "{1}"'
+                    .format(param, cls.__name__))
+
+    """
+    Decide if it's appropriate to log given a context, and update the context accordingly.
+    """
+    @staticmethod
+    def log_condition(context):
+        return True
+
+
+class Once(LoggingFilter):
+    """Ignore all log calls except the first one."""
+
+    params = {
+        'once': None,
+    }
+
+    @classmethod
+    def initialize_context(cls, context, **kwargs):
+        super(cls, cls).initialize_context(context, **kwargs)
```

Comment:
You can't use `cls` as the fist argument but need to use `Once`. Otherwise if another class subclasses from `Once` this wouldn't work.

Same below.

Reply:
argh, thanks. I corrected this in 0c7916c49d178259f1602b5847c96e1d7a509154 but can remove the arguments if we prefer easy copy-paste over explicitness